doc/designs/thread-api.md

```diff
@@ -0,0 +1,304 @@
+ *
+ * If the argument is 0, thread pooling is disabled. OpenSSL will not create any
+ * threads and existing threads in the thread pool will be torn down.
+ *
+ * Returns 1 on success and 0 on failure. Returns failure if OpenSSL-managed
+ * thread pooling is not supported (for example, if it is not supported on the
+ * current platform, or because OpenSSL is not built with the necessary
+ * support).
+ */
+int OSSL_set_max_threads(uint32_t max_threads);
+
+/*
+ * Get the maximum number of threads currently allowed to be used by the
+ * thread pool. If thread pooling is disabled or not available, returns 0.
+ */
+uint32_t OSSL_get_max_threads(void);
+```
+
+The maximum thread count is a limit, not a target. Threads will not be spawned
+unless (and until) there is demand.
+
+Custom Thread Spawner Method
+----------------------------
+
+If an application wishes to control thread spawning, it may set a custom set of
+thread method functions. OpenSSL will use these to spawn and join threads.
+
+```c
+typedef struct ossl_thread_cookie_st OSSL_THREAD_COOKIE;
+
+typedef struct ossl_thread_method_st {
+    /*
+     * Spawns thread, calling cb with arg. An opaque cookie representing the
+     * thread is written to *thread_cookie.
+     *
+     * Returns 1 on success or 0 on failure.
+     */
+    int (*spawn_thread)(void (*cb)(void *arg), void *arg,
+                        OSSL_THREAD_COOKIE **thread_cookie);
+
+    /*
+     * Wait for a thread to exit.
+     */
+    void (*join_thread)(OSSL_THREAD_COOKIE *thread_cookie);
+} OSSL_THREAD_METHOD;
+
+/*
+ * Call OSSL_set_thread_method(&method, sizeof(method)) to set a new thread
+ * method. This must be called before OpenSSL creates any threads (e.g. before
+ * the call to `OSSL_set_max_threads`), else it fails.
+ *
+ * The length argument allows ABI compatibility to be preserved if methods are
+ * added to the OSSL_THREAD_METHOD structure in future.
+ */
+int OSSL_set_thread_method(OSSL_THREAD_METHOD *method, size_t len);
+```
+
+OpenSSL uses the custom thread spawner method in accordance with the thread
+limit set using the default API already introduced. However if necessary, an
+application can also control thread creation by causing the `spawn_thread`
+method to fail.
+
+Note that this gives the application control over thread spawning and joining,
+but not other threading functionality like mutexes or condition variables.
+The rationale for this is:
+
+- Thread creation and joining requires OS interaction (syscalls);
+- Mutexes or condition variables do not necessarily require OS interaction;
+- It is comparatively more likely for applications to want to intercept
+  thread management than synchronisation primitives;
+- Making synchronisation primitives polymorphic would be overkill
+  and add unnecessary overhead;
+- We already use synchronisation primitives (e.g. pthread rwmutex)
+  without providing options to the application.
+
+This method can be dropped from MVP if it is deemed unnecessary.
```

Comment:
IMHO we really do not want this option in MVP.

Reply:
Agreed, let's not expose this publicly in MVP if we can help it.  Let it mature a bit first.